src/bson.rs

```diff
@@ -973,76 +970,78 @@ impl Timestamp {
+///
+/// Is is recommended to use a [`chrono::DateTime`] for date operations
+/// and to convert it to/from a [`crate::DateTime`] via the `From`/`Into` implementations.
+///
+/// ```
+/// use chrono::prelude::*;
+/// # fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
+/// let chrono_dt: chrono::DateTime<Utc> = "2014-11-28T12:00:09Z".parse()?;
+/// let bson_dt: bson::DateTime = chrono_dt.into();
+/// let back_to_chrono: chrono::DateTime<Utc> = bson_dt.into();
+/// # Ok(())
+/// # }
+/// ```
 ///
-/// Just a helper for convenience
+/// This type differs from [`chrono::DateTime`] in that it serializes to and deserializes from a
+/// BSON datetime rather than an ISO-8601 formatted string. This means that it will serialize to and
+/// deserialize from an extended JSON object in non-BSON formats. To serialize a
+/// [`chrono::DateTime`] as a BSON datetime, you can use
+/// [`serde_helpers::chrono_0_4_datetime_as_bson_datetime`].
 ///
 /// ```rust
 /// use serde::{Serialize, Deserialize};
-/// use bson::DateTime;
 ///
 /// #[derive(Serialize, Deserialize)]
 /// struct Foo {
-///     date_time: DateTime,
+///     // serializes as a BSON datetime.
+///     date_time: bson::DateTime,
+///
+///     // serializes as an ISO-8601 string.
+///     chrono_datetime: chrono::DateTime<chrono::Utc>,
+///
+///     // serializes as a BSON datetime.
+///     #[serde(with = "bson::serde_helpers::chrono_0_4_datetime_as_bson_datetime")]
+///     chrono_as_bson: chrono::DateTime<chrono::Utc>,
 /// }
 /// ```
 #[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Copy, Clone)]
-pub struct DateTime(pub chrono::DateTime<Utc>);
-
-impl DateTime {
-    pub(crate) fn from_i64(date: i64) -> Self {
-        let mut num_secs = date / 1000;
-        let mut num_millis = date % 1000;
-
-        // The chrono API only lets us create a DateTime with an i64 number of seconds
-        // and a u32 number of nanoseconds. In the case of a negative timestamp, this
-        // means that we need to turn the negative fractional part into a positive and
-        // shift the number of seconds down. For example:
-        //
-        //     date       = -4300 ms
-        //     num_secs   = date / 1000 = -4300 / 1000 = -4
-        //     num_millis = date % 1000 = -4300 % 1000 = -300
-        //
-        // Since num_millis is less than 0:
-        //     num_secs   = num_secs -1 = -4 - 1 = -5
-        //     num_millis = num_nanos + 1000 = -300 + 1000 = 700
-        //
-        // Instead of -4 seconds and -300 milliseconds, we now have -5 seconds and +700
-        // milliseconds, which expresses the same timestamp, but in a way we can create
-        // a DateTime with.
-        if num_millis < 0 {
-            num_secs -= 1;
-            num_millis += 1000;
-        };
+pub struct DateTime(chrono::DateTime<Utc>);
 
-        Utc.timestamp(num_secs, num_millis as u32 * 1_000_000)
-            .into()
+impl crate::DateTime {
+    pub(crate) fn from_millis(date: i64) -> Self {
```

Comment:
I thought about making this public, but apparently too-large values here can cause panics (RUST-799), so I just left it private for now until we decide how / if we want to tackle this publicly.